### PR TITLE
Flesh out get recogniser step of pipeline

### PR DESCRIPTION
**Pull request by gabechu**
### Description
Fill up stub method of get_recogniser in pipeline. It fetches the model class that we'd like to instantiate.

In summary, now do something like this you can have an instantiated recogniser:
```
# registry will be auto built when you import
from pii_recognition import registry

all_predefined_recognisers = registry.recogniser
recogniser_1 = registry.recogniser(name: str="Recogniser_1", config_of_recogniser_1: Dict)
```

### Checklist
- [x] Added **tests** for changed code.

### Todos
- [x] Config argument as optional
- [x] Verify if can pass `**Dict` as an argument for `pakkr` (solution use Dict instead no asterisks)
- [x] Fix `get_recogniser` test, which is the first step of the pipeline

### CODE REVIEW

File: pii_recognition/evaluation/pipeline_test.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    "pii_recognition.evaluation.pipeline.recogniser_registry",
+    new={"fake_model": "fake_value"},
+)
+def test_get_recogniser():
+    actual = get_recogniser("fake_model")
+    assert actual == "fake_value"
+
+    with raises(KeyError) as err:
+        get_recogniser("non_exist_model")
+    assert str(err.value) == "'non_exist_model'"
```
Comment (kerrychu-zen) [inline pii_recognition/evaluation/pipeline_test.py:18]: same here. more explicit error message pls. `recognizer requested is not registered`

File: pii_recognition/registers/recogniser_registry_test.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+from .recogniser_registry import RecogniserRegistry
+
+
+def test_RecogniserRegistry():
+    actual = RecogniserRegistry()
+    assert isinstance(actual, dict)
+    assert len(actual.keys()) > 0
```
Comment (kerrychu-zen) [inline pii_recognition/registers/recogniser_registry_test.py:7]: why not test the exact key length?

Comment (gabechu) [reply pii_recognition/registers/recogniser_registry_test.py:7]: I thought about that and the consideration is that as the project moves, we may have more and more predefined recognisers, regex recognisers. Whenever adding a recogniser, test has to be updated accordingly, a lot of manual changes involved.

Here the test is to check if the registry is empty and my assumption is that if we want to use a registry, there must be something in it. So far, this assumption is still true. And if you have any other thoughts on more efficient implementation of registry as well as testing, drop a line or two.

File: pii_recognition/evaluation/pipeline_test.py (hunk 1 above)
Comment (gabechu) [reply pii_recognition/evaluation/pipeline_test.py:18]: Current setting is to use registry as a dictionary, i.e.,
```
{"model_1": model_1_imp_class, "model_2": model_2_imp_class}
{"tokeniser_1": tokeniser_1_imp, "tokeniser_2": tokeniser_2_imp}
```
and all error messages are inherited from `dict` as well, the full message of a dictionary error you will see in terminal is `KeyError: model_3` or `KeyError: tokeniser_3`.

To implement more meaningful changes have to do something below
```
class Regsitry( meta_class=dict ):
    def __getitem__( self, key ):
        try:
            return super( ).__getitem__( key )
        except KeyError as e:
            raise MyErrorMessage( e.message )
```
I guess the question is that do users expect a `KeyError` message when using a dict-like object, or do users actually are expecting a meaningful custom error instead.

**Comment by kerrychu-zen** [comment]
I think it is great to explore using this pattern to hold all the tokenizer objects in one place. The benefits are that the registered key-value pair can be used repetitively and globally.

But maybe it's better to stop and think about is this pattern suitable for the problem we try to solve. The registry pattern is obviously not easy to write the unit test and when it comes to writing pipelines, for each task I assume you just need to instantiate a tokeniser it once. Hence, I think a simple dictionary should do just fine. The registry pattern seems to be an overkill.

**Comment by gabechu** [comment]
@kerrychu-zen I feel I didn’t make it clear about the registry which being introduced in the last PR. The only difference between the registry and simple dict is whether the object has been type annotated. 
```
# simple dict
recogniser: Dict = {"recogniser_1": recogniser_1}
reader: Dict = {"reader_1": reader_1}

# registry
recogniser: Registry[Recogniser] = {"recogniser_1": recogniser_1}
reader: Registry[Reader] = {"reader_1": reader_1}
```
In the case of simple dict, there’s no concept of typing, everything is treat equally. But with registry, it is still a dict but it is a typed dict.

Another question may be asked is that why we want things typed, it’s all a matter of preference. I found with typing, it prevent dumb errors as well as helps documentation. Regarding to unittest when building pipelines, it would make no difference because type hinting is not for runtime checking which won't affect unittest, and all we need to do is to mock out registry when testing:
```
mock_registry = {"recogniser_1": recogniser_1}
```

**Comment by kerrychu-zen** [comment]
I see your concerns. Have you checked [TypedDict](https://docs.python.org/3/library/typing.html#typing.TypedDict)? @gabechu

**Comment by gabechu** [comment]
I just read up on the document, that look cools. It is a new feature that is available in 3.8 I may need to bump the version to use this feature (but not sure if pytorch has supports for 3.8 yet). Yes, TypedDict is an equivalent concept. I believe we can have an implementation like this with TypedDict

### TypedDict
```
tokeniser_registry = TypedDict('Tokeniser', {'Tokeniser_1': Tokeniser_1, 'Tokeniser_2': Tokeniser_2})
# a separate function to instantiate an instance
create_instance(tokeniser_registry, name="Tokeniser_1")
```
### current implementation
```
class Registry(Generics[T]):
    def add_item(item): ...
    def create_instance(name): ...

# need a base class Tokeniser
tokeniser_registry = Registry[Tokeniser]()
tokeniser_registry.add_item(Tokeniser_1)
tokeniser_registry.create_instance("Tokeniser_1")
```
it depends on whether you want the encapsulation.

**Comment by kerrychu-zen** [comment]
I think `machine-learning-repo` is still using python `3.7.4` better not bumping this repo to keep it consistent (in case you come back at some point and we need to migrate this repo to `machine-learning-research`). I forgot `TypedDict` is a 3.8 thing.

Just brainstorming to see what's a better fit for the problem we are trying to solve. I still feel a bit off with Registry but happy to take on the journey with you and see how the registry serves us.

I have a suggestion: maybe push stubbed architecture PR first in the future then people can have an overall idea and it's easier to give feedback on approaches. @gabechu what do you think?

File: pii_recognition/registration/bind.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+        recognisers.first_letter_uppercase_recogniser.FirstLetterUppercaseRecogniser,
+        recognisers.flair_recogniser.FlairRecogniser,
+        recognisers.spacy_recogniser.SpacyRecogniser,
+        recognisers.stanza_recogniser.StanzaRecogniser,
+    ]
+
+
+def _run():
+    for item in _get_default_recognisers():
+        registry.recogniser.add_item(item)
```
Comment (kerrychu-zen) [inline pii_recognition/registration/bind.py:17]: I think we need a more meaningful name for this function. Maybe `_register_recognisers`?
other than this I think this PR is good to go.

Comment (gabechu) [reply pii_recognition/registration/bind.py:17]: That a much better name! I will update it.

**Comment by gabechu** [comment]
I totally agree. The registry class has evolved quite a lot from the initial design. The goal of [the first iteration](https://github.com/gabechu/pii_recognition/blob/master/pii_recognition/experiments/python_crf.py) is done it quick for evaluation results without consider too much on scale or design. When executing in pipeline, this becomes troublesome -- add design without breaking existing code. As you suggested, for this kind of structure changes, it is good to use stubs to help outline the direction for people pondering such a shift.

**Comment by gabechu** [comment]
I totally agree. The registry class has evolved quite a lot from the initial design. The goal of[ the first iteration](https://github.com/gabechu/pii_recognition/blob/master/pii_recognition/experiments/python_crf.py) is done it quick for evaluation results without considering too much of scaling or design. When executing in pipeline for another iteration, this becomes troublesome -- adding design without breaking existing code. As you suggested, for this kind of structure changes, it is good to use stubs to help outline the direction that people can ponder on such a shift.

**Pull request merged**